backend/cfg/llvmize.ml

```diff
@@ -0,0 +1,574 @@
+[@@@ocaml.warning "+a-30-40-41-42"]
+
+module CL = Cfg_with_layout
+module DLL = Oxcaml_utils.Doubly_linked_list
+
+type error = Asm_generation of (string * int)
+
+exception Error of error
+
+module Ident : sig
+  (** Unnamed LLVM identifiers that start with "%".  This includes
+      basic blocks, unnamed function parameters, and temporaries
+      (virtual registers).  *)
+  type t
+
+  val print : Format.formatter -> t -> unit
+
+  module Gen : sig
+    (** per-function counter for generating identifiers *)
+    type ident = t
+
+    type t
+
+    val create : unit -> t
+
+    val get_fresh : t -> ident
+  end
+end = struct
+  type t = int
+
+  let print fmt t = Format.fprintf fmt "%d" t
+
+  module Gen = struct
+    type ident = t
+
+    type nonrec t = { mutable next : ident }
+
+    (* Local identifiers are only valid within function scope, so we can reset
+       it to 0 every time *)
+    (* let create () = { next = Label.to_int (Cmm.new_label ()) } *)
+    let create () = { next = 0 }
+
+    let get_fresh t =
+      let res = t.next in
+      t.next <- succ res;
+      res
+  end
+end
+
+module Llvm_typ = struct
+  (** Type representing LLVM types *)
+  type t =
+    | Int of int  (** Arbitrary-sized (in bits) integer *)
+    | Ptr
+    | Struct of t list
+
+  let of_machtyp_component (c : Cmm.machtype_component) =
+    match c with
+    | Val | Addr | Int ->
+      Int 64
+      (* CFG allows vals to be assigned to ints and vice versa, so we do this to
+         make LLVM happy for now. *)
+    | Float | Vec128 | Vec256 | Vec512 | Float32 | Valx2 ->
+      Misc.fatal_error "Not implemented [Llvm_typ.of_machtyp_component]"
+
+  let rec pp_t ppf t =
+    let open Format in
+    match t with
+    | Int n -> fprintf ppf "i%d" n
+    | Ptr -> fprintf ppf "ptr"
+    | Struct typs ->
+      fprintf ppf "{ %a }"
+        (pp_print_list ~pp_sep:(fun ppf () -> fprintf ppf ", ") pp_t)
+        typs
+
+  let to_string t = Format.asprintf "%a" pp_t t
+end
+
+type fun_info =
+  { ident_gen : Ident.Gen.t;
+    reg2ident : Ident.t Reg.Tbl.t;  (** Map register's stamp to identifier  *)
+    label2ident : Ident.t Label.Tbl.t;
+        (** Map label to identifier. Avoid clashes between pre-existing Cfg labels and unnamed identifiers. *)
+    mutable arg_regs : Reg.Set.t;
+        (** Registers mentioned in the argument list of the function. *)
+    mutable temp_regs : Reg.Set.t
+        (** Temporary registers mentioned in the body of the function. *)
+  }
+
+type t =
+  { llvmir_filename : string;
+    oc : Out_channel.t;
+    ppf : Format.formatter;
+    asm_filename : string;
+    current_fun_info : fun_info ref; (* should be reset for every function *)
+    data : Cmm.data_item list ref
+  }
+
+let create_fun_info () =
+  { ident_gen = Ident.Gen.create ();
+    reg2ident =
+      Reg.Tbl.create 37 (* CR yusumez: change this to be more reasonable *);
+    label2ident = Label.Tbl.create 37;
+    arg_regs = Reg.Set.empty;
+    temp_regs = Reg.Set.empty
```

Comment:
I don't think we need to keep `temp_regs` after `alloca_regs`, and if so, `collect_regs` can be done as part of `alloca_regs`: just alloca them as soon as you see them. Then, avoid overwriting previously `alloca`d identifier for fun_arg but checking if something is arg before looking it up.